File: lintcode/python/0014_first_position_of_target.py
#!/usr/bin/python -t

# binary search solution, time O(logn) space O(1)

class Solution:
    """
    @param nums: The integer array.
    @param target: Target to find.
    @return: The first position of target. Position starts from 0.
    """
    def binarySearch(self, nums, target):
        # write your code here
        # find lower_bound

        n = len(nums)
        l = 0
        r = n-1
        
        while l < r:
            mid = (l+r)/2
            
            if nums[mid] < target:
                l = mid+1
            else:
                r = mid
                
        return -1 if nums[l] != target else l

File: lintcode/golang/0014_first_position_of_target.go
/**
 * @param nums: The integer array.
 * @param target: Target to find.
 * @return: The first position of target. Position starts from 0.
 */
 
//import "fmt"

func binarySearch (nums []int, target int) int {
    // write your code here
    left := 0
    right := len(nums)-1
    
    for left + 1 < right {
       mid := (left+right)/2
       if nums[mid] >= target {
           right = mid
       } else {
           left = mid
       }
    }
    
    //fmt.Println(left, right)
    if nums[left] == target {
        return left
    }
    if nums[right] == target {
        return right
    }
    
    return -1
}

